From f0bdc99cdbe140902c6ddda9c09787b3dea8b50a Mon Sep 17 00:00:00 2001
From: Jacky Bai <ping.bai@nxp.com>
Date: Wed, 12 Feb 2020 13:35:57 +0800
Subject: [PATCH 22/37] MLK-23305 cpufreq: imx8mq: Fix the incorrect voltage
 scaling

commit f6106f0d822348658f33d1047f59d63e77133d3f from
https://source.codeaurora.org/external/imx/linux-imx.git

When the policy max frequency is changed due to thermal cooling,
the dc regulator's voltage will mismatch with the OPP table,
so use new_freq & old_freq to determine the voltage each time
the new frequency is selected.

Signed-off-by: Jacky Bai <ping.bai@nxp.com>
Reviewed-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/imx8mq-cpufreq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/cpufreq/imx8mq-cpufreq.c b/drivers/cpufreq/imx8mq-cpufreq.c
index 615f05d36f0a..fcb35ad07c2f 100644
--- a/drivers/cpufreq/imx8mq-cpufreq.c
+++ b/drivers/cpufreq/imx8mq-cpufreq.c
@@ -57,7 +57,7 @@ static int imx8mq_set_target(struct cpufreq_policy *policy, unsigned int index)
 	dev_dbg(cpu_dev, "%u MHz --> %u MHz\n",
 		old_freq / 1000, new_freq / 1000);
 
-	if (new_freq == policy->max) {
+	if (new_freq > old_freq) {
 		if (!IS_ERR(dc_reg)) {
 			ret = regulator_set_voltage_tol(dc_reg, DC_VOLTAGE_MAX, 0);
 			if (ret) {
@@ -81,7 +81,7 @@ static int imx8mq_set_target(struct cpufreq_policy *policy, unsigned int index)
 	clk_set_rate(arm_pll_clk, new_freq * 1000);
 	clk_set_parent(arm_a53_src_clk, arm_pll_out_clk);
 
-	if (old_freq == policy->max) {
+	if (new_freq < old_freq) {
 		if (!IS_ERR(dc_reg)) {
 			ret = regulator_set_voltage_tol(dc_reg, DC_VOLTAGE_MIN, 0);
 			if (ret) {
-- 
2.17.1

